lib/charms/loki_k8s/v0/loki_push_api.py

```diff
@@ -977,48 +998,45 @@ def _on_logging_relation_changed(self, event):
             loki_push_api_alert_rules_error: This event is emitted when an invalid alert rules
                 file is encountered or if `alert_rules_path` is empty.
         """
-        if not self._charm.unit.is_leader():
-            return
-
-        if event.unit is None:
-            # Workaround: Seems this is a Juju bug that sends event.unit == None
-            # Remove this if when this issue is closed:
-            # https://github.com/canonical/loki-operator/issues/3
-            return
-
-        self._store_loki_push_api(event)
-
-        alert_groups, invalid_files = load_alert_rules_from_dir(
-            self._alert_rules_path,
-            self.topology,
-            recursive=False,
-            allow_free_standing=self.allow_free_standing_rules,
-        )
-        alert_rules_error_message = self._check_alert_rules(alert_groups, invalid_files)
+        if isinstance(event, RelationEvent):
+            self._process_logging_relation_changed(event.relation)
+        else:
+            # Upgrade event or other charm-level event
+            for relation in self._charm.model.relations[self._relation_name]:
+                self._process_logging_relation_changed(relation)
+
+    def _process_logging_relation_changed(self, relation: Relation):
+        if loki_push_api_data := relation.data[relation.app].get("loki_push_api"):
+            self._stored.loki_push_api[relation.id] = json.loads(loki_push_api_data)
```

Comment:
Not sure why duplicate it into `_stored` if it is readily available in relation data.